File: python/src/apx_agent/_managed_mcp.py
"""Databricks Managed MCP integration — generate URLs and client configs.

Databricks Managed MCP (Public Preview, Jan 2026) exposes Unity Catalog
functions, Genie spaces, and Vector Search indexes as MCP servers, hosted
by the platform. Anything an apx-agent declares as a Mosaic AI resource
becomes a Managed MCP endpoint automatically — there's no per-agent MCP
server to deploy, no app.yaml to maintain, no naming conventions for
discovery.

This module provides two helpers:

  * ``managed_mcp_urls(agent, workspace_host)`` — walks the agent's
    declared ``ResourceSpec`` list and emits the Managed MCP URL for
    each resource, paired with the OAuth scope it needs.

  * ``managed_mcp_client_config(endpoints, name=...)`` — produces the
    MCP server config snippet ready to drop into Claude Desktop,
    Cursor, or any client that speaks the standard ``mcpServers``
    config shape.

URL patterns (from the Databricks managed MCP docs):
  * UC function:   ``/api/2.0/mcp/functions/{catalog}/{schema}/{function_name}``
  * Genie space:   ``/api/2.0/mcp/genie/{genie_space_id}``
  * Vector Search: ``/api/2.0/mcp/vector-search/{catalog}/{schema}/{index_name}``

Resources outside those three kinds (serving endpoints, SQL warehouses,
UC tables) don't have a corresponding Managed MCP server today — the
walker reports them with ``url=None`` and ``unsupported=True`` so the
caller can decide how to surface them.
"""

from __future__ import annotations

import logging
from dataclasses import dataclass
from typing import TYPE_CHECKING

from ._resources import collect_resource_specs

if TYPE_CHECKING:
    from ._agents import BaseAgent

logger = logging.getLogger(__name__)


# ---------------------------------------------------------------------------
# Result type
# ---------------------------------------------------------------------------


@dataclass(frozen=True)
class ManagedMCPEndpoint:
    """A single Databricks Managed MCP endpoint.

    Attributes:
        kind: Resource kind from the underlying ``ResourceSpec`` —
            ``"uc_function"``, ``"genie_space"``,
            ``"vector_search_index"``, or one of the unsupported kinds
            (``"serving_endpoint"``, ``"sql_warehouse"``, ``"uc_table"``).
        identifier: The resource identifier (full UC name, space id, etc.).
        url: The fully-qualified Managed MCP URL. ``None`` for unsupported
            kinds.
        oauth_scope: The OAuth scope a client must request to call the
            endpoint. ``None`` for unsupported kinds.
        unsupported: ``True`` when this resource kind has no Managed MCP
            equivalent (the caller may want to skip it or surface it
            separately).
    """

    kind: str
    identifier: str
    url: str | None
    oauth_scope: str | None
    unsupported: bool = False


# ---------------------------------------------------------------------------
# Mapping table
# ---------------------------------------------------------------------------


# kind -> (path template, oauth scope). Path templates use {0}, {1}, {2}
# placeholders filled from the dotted identifier.
_KIND_TO_PATH: dict[str, tuple[str, str]] = {
    "uc_function": ("/api/2.0/mcp/functions/{0}/{1}/{2}", "unity-catalog"),
    "genie_space": ("/api/2.0/mcp/genie/{0}", "genie"),
    "vector_search_index": ("/api/2.0/mcp/vector-search/{0}/{1}/{2}", "vector-search"),
}


def _normalise_host(workspace_host: str) -> str:
    """Strip trailing slash and ensure the host has a scheme."""
    host = workspace_host.rstrip("/")
    if not host.startswith(("http://", "https://")):
        host = f"https://{host}"
    return host


def _build_url(kind: str, identifier: str, workspace_host: str) -> tuple[str | None, str | None]:
    """Return ``(url, oauth_scope)`` for the kind/identifier pair, or ``(None, None)``."""
    template_scope = _KIND_TO_PATH.get(kind)
    if template_scope is None:
        return None, None
    template, scope = template_scope

    if kind == "genie_space":
        return f"{workspace_host}{template.format(identifier)}", scope

    # UC function / vector search use three-part identifiers
    parts = identifier.split(".")
    if len(parts) != 3:
        logger.warning(
            "Managed MCP URL for %s expects a three-part identifier; got %r — skipping.",
            kind, identifier,
        )
        return None, None
    return f"{workspace_host}{template.format(*parts)}", scope


# ---------------------------------------------------------------------------
# Public API
# ---------------------------------------------------------------------------


def managed_mcp_urls(
    agent: "BaseAgent",
    *,
    workspace_host: str,
    model: str | None = None,
) -> list[ManagedMCPEndpoint]:
    """Return the Managed MCP endpoints for every resource the agent declares.

    Args:
        agent: The apx-agent ``BaseAgent`` to walk.
        workspace_host: Databricks workspace host (e.g.
            ``"https://my-workspace.cloud.databricks.com"`` — scheme
            optional, trailing slash optional).
        model: Optional model serving endpoint name to include in the
            walk. Passed through to ``collect_resource_specs``; usually
            irrelevant for Managed MCP since model endpoints aren't
            exposed as MCP servers, but kept for symmetry with
            ``log_agent``.

    Returns:
        One ``ManagedMCPEndpoint`` per declared resource. Endpoints whose
        kind has no Managed MCP equivalent are returned with
        ``unsupported=True`` and ``url=None`` so the caller can choose to
        filter or surface them separately. Ordering preserves the
        walker's first-occurrence order.
    """
    host = _normalise_host(workspace_host)
    specs = collect_resource_specs(agent, model=model)

    endpoints: list[ManagedMCPEndpoint] = []
    for spec in specs:
        url, scope = _build_url(spec.kind, spec.identifier, host)
        if url is None:
            endpoints.append(ManagedMCPEndpoint(
                kind=spec.kind,
                identifier=spec.identifier,
                url=None,
                oauth_scope=None,
                unsupported=True,
            ))
        else:
            endpoints.append(ManagedMCPEndpoint(
                kind=spec.kind,
                identifier=spec.identifier,
                url=url,
                oauth_scope=scope,
                unsupported=False,
            ))
    return endpoints


def managed_mcp_client_config(
    endpoints: list[ManagedMCPEndpoint],
    *,
    name: str = "databricks",
    include_unsupported: bool = False,
) -> dict:
    """Produce an MCP client config snippet from a list of endpoints.

    Returns a dict in the standard ``mcpServers`` shape understood by
    Claude Desktop, Cursor, and other MCP-aware clients. Drop the
    returned dict into the client's config file (``claude_desktop_config.json``,
    ``~/.cursor/mcp.json``, etc.) under the ``mcpServers`` key.

    Each endpoint becomes one server entry. The key is derived from the
    kind and identifier — e.g. ``"databricks.uc_function.main.tools.classify_intent"``
    — so multiple endpoints from one agent don't collide.

    Args:
        endpoints: Output of ``managed_mcp_urls(...)``.
        name: Prefix applied to every server key. Defaults to
            ``"databricks"``. Override to namespace multiple agents
            cleanly in the same client config.
        include_unsupported: When ``False`` (default), endpoints with
            ``unsupported=True`` are skipped. Set ``True`` if you want
            them in the output anyway (e.g. as a TODO list).

    Returns:
        A dict of shape::

            {
              "mcpServers": {
                "databricks.uc_function.main.tools.classify_intent": {
                  "type": "http",
                  "url": "https://.../api/2.0/mcp/functions/main/tools/classify_intent",
                  "oauth_scope": "unity-catalog"
                },
                ...
              }
            }
    """
    servers: dict[str, dict] = {}
    for ep in endpoints:
        if ep.unsupported and not include_unsupported:
            continue
        key = f"{name}.{ep.kind}.{ep.identifier}"
        entry: dict = {"type": "http", "url": ep.url}
        if ep.oauth_scope:
            entry["oauth_scope"] = ep.oauth_scope
        if ep.unsupported:
            entry["unsupported"] = True
        servers[key] = entry
    return {"mcpServers": servers}

File: typescript/src/managed-mcp.ts
/**
 * Databricks Managed MCP integration — generate URLs and client configs.
 *
 * Databricks Managed MCP (Public Preview, Jan 2026) exposes Unity Catalog
 * functions, Genie spaces, and Vector Search indexes as MCP servers, hosted
 * by the platform. Anything an apx-agent declares as a Mosaic AI resource
 * becomes a Managed MCP endpoint automatically — there's no per-agent MCP
 * server to deploy, no app.yaml to maintain, no naming conventions for
 * discovery.
 *
 * This module provides two helpers:
 *
 *   - {@link managedMcpUrls} — takes an explicit array of declared
 *     `ResourceSpec`s and emits the Managed MCP URL for each resource,
 *     paired with the OAuth scope it needs.
 *
 *   - {@link managedMcpClientConfig} — produces the MCP server config
 *     snippet ready to drop into Claude Desktop, Cursor, or any client
 *     that speaks the standard `mcpServers` config shape.
 *
 * URL patterns (from the Databricks managed MCP docs):
 *   - UC function:   `/api/2.0/mcp/functions/{catalog}/{schema}/{function_name}`
 *   - Genie space:   `/api/2.0/mcp/genie/{genie_space_id}`
 *   - Vector Search: `/api/2.0/mcp/vector-search/{catalog}/{schema}/{index_name}`
 *
 * Resources outside those three kinds (serving endpoints, SQL warehouses,
 * UC tables) don't have a corresponding Managed MCP server today — those
 * are reported with `url=null` and `unsupported=true` so the caller can
 * decide how to surface them.
 */

import { z } from 'zod';
import type { ResourceKind, ResourceSpec } from './resources.js';

// ---------------------------------------------------------------------------
// Types
// ---------------------------------------------------------------------------

/**
 * Re-export the canonical resource types from `./resources.js`. The strict
 * `ResourceKind` union and `ResourceSpec` interface live there; this module
 * re-exports them so existing callers importing from `managed-mcp.js`
 * continue to work.
 */
export type { ResourceKind, ResourceSpec } from './resources.js';

/**
 * A single Databricks Managed MCP endpoint.
 *
 * - `kind` — resource kind from the underlying `ResourceSpec`.
 * - `identifier` — the resource identifier (full UC name, space id, etc.).
 * - `url` — fully-qualified Managed MCP URL. `null` for unsupported kinds.
 * - `oauthScope` — OAuth scope a client must request. `null` for unsupported.
 * - `unsupported` — `true` when this kind has no Managed MCP equivalent.
 *
 * Instances are frozen so callers can rely on referential immutability.
 */
export interface ManagedMCPEndpoint {
  readonly kind: ResourceKind;
  readonly identifier: string;
  readonly url: string | null;
  readonly oauthScope: string | null;
  readonly unsupported: boolean;
}

// ---------------------------------------------------------------------------
// Mapping table
// ---------------------------------------------------------------------------

/**
 * Kind → (path template, oauth scope). The template uses `{0}`, `{1}`,
 * `{2}` placeholders filled from the dotted identifier (or just `{0}` for
 * single-part identifiers like genie spaces).
 */
const KIND_TO_PATH: Record<string, { template: string; scope: string }> = {
  uc_function: {
    template: '/api/2.0/mcp/functions/{0}/{1}/{2}',
    scope: 'unity-catalog',
  },
  genie_space: {
    template: '/api/2.0/mcp/genie/{0}',
    scope: 'genie',
  },
  vector_search_index: {
    template: '/api/2.0/mcp/vector-search/{0}/{1}/{2}',
    scope: 'vector-search',
  },
};

function fillTemplate(template: string, parts: string[]): string {
  return template.replace(/\{(\d+)\}/g, (_, idx) => parts[Number(idx)] ?? '');
}

function normaliseHost(workspaceHost: string): string {
  let host = workspaceHost.replace(/\/+$/, '');
  if (!/^https?:\/\//i.test(host)) {
    host = `https://${host}`;
  }
  return host;
}

function buildUrl(
  kind: string,
  identifier: string,
  workspaceHost: string,
): { url: string | null; oauthScope: string | null } {
  const mapping = KIND_TO_PATH[kind];
  if (!mapping) {
    return { url: null, oauthScope: null };
  }

  if (kind === 'genie_space') {
    const path = fillTemplate(mapping.template, [identifier]);
    return { url: `${workspaceHost}${path}`, oauthScope: mapping.scope };
  }

  // UC function / vector search use three-part identifiers
  const parts = identifier.split('.');
  if (parts.length !== 3) {
    console.warn(
      `[managed-mcp] Managed MCP URL for ${kind} expects a three-part identifier; got ${JSON.stringify(identifier)} — skipping.`,
    );
    return { url: null, oauthScope: null };
  }
  const path = fillTemplate(mapping.template, parts);
  return { url: `${workspaceHost}${path}`, oauthScope: mapping.scope };
}

// ---------------------------------------------------------------------------
// Public API
// ---------------------------------------------------------------------------

const ManagedMcpUrlsOptionsSchema = z.object({
  resourceSpecs: z.array(
    z.object({
      kind: z.string(),
      identifier: z.string(),
    }),
  ),
  workspaceHost: z.string().min(1),
  model: z.string().optional(),
});

export interface ManagedMcpUrlsOptions {
  /** Resource specs declared by the agent (or assembled by the caller). */
  resourceSpecs: ResourceSpec[];
  /**
   * Databricks workspace host (e.g. `https://my-workspace.cloud.databricks.com`).
   * Scheme and trailing slash are optional — normalised internally.
   */
  workspaceHost: string;
  /**
   * Optional model serving endpoint name. When supplied, a
   * `{kind: 'serving_endpoint', identifier: model}` spec is prepended to
   * the input — mirrors the Python helper's `model=` kwarg.
   */
  model?: string;
}

/**
 * Return the Managed MCP endpoints for an explicit list of `ResourceSpec`s.
 *
 * Endpoints whose kind has no Managed MCP equivalent are returned with
 * `unsupported=true` and `url=null` so the caller can choose to filter or
 * surface them separately. Ordering preserves input order (with the model
 * spec prepended, when supplied).
 */
export function managedMcpUrls(options: ManagedMcpUrlsOptions): ManagedMCPEndpoint[] {
  const parsed = ManagedMcpUrlsOptionsSchema.parse(options);
  const host = normaliseHost(parsed.workspaceHost);

  // Use the loose shape internally — buildUrl tolerates unknown kinds and
  // returns null/null when no Managed MCP path applies, so unsupported
  // kinds flow through naturally.
  const specs: { kind: string; identifier: string }[] = parsed.model
    ? [{ kind: 'serving_endpoint', identifier: parsed.model }, ...parsed.resourceSpecs]
    : [...parsed.resourceSpecs];

  return specs.map((spec) => {
    const { url, oauthScope } = buildUrl(spec.kind, spec.identifier, host);
    if (url === null) {
      return Object.freeze({
        kind: spec.kind as ResourceKind,
        identifier: spec.identifier,
        url: null,
        oauthScope: null,
        unsupported: true,
      }) as ManagedMCPEndpoint;
    }
    return Object.freeze({
      kind: spec.kind as ResourceKind,
      identifier: spec.identifier,
      url,
      oauthScope,
      unsupported: false,
    }) as ManagedMCPEndpoint;
  });
}

const ManagedMcpClientConfigOptionsSchema = z.object({
  endpoints: z.array(
    z.object({
      kind: z.string(),
      identifier: z.string(),
      url: z.string().nullable(),
      oauthScope: z.string().nullable(),
      unsupported: z.boolean(),
    }),
  ),
  name: z.string().min(1).optional(),
  includeUnsupported: z.boolean().optional(),
});

export interface ManagedMcpClientConfigOptions {
  /** Output of {@link managedMcpUrls}. */
  endpoints: ManagedMCPEndpoint[];
  /**
   * Prefix applied to every server key. Defaults to `"databricks"`.
   * Override to namespace multiple agents in the same client config.
   */
  name?: string;
  /**
   * When `false` (default), endpoints with `unsupported=true` are skipped.
   * Set `true` to include them anyway (e.g. as a TODO list).
   */
  includeUnsupported?: boolean;
}

export interface McpServerEntry {
  type: 'http';
  url: string | null;
  oauth_scope?: string;
  unsupported?: boolean;
}

export interface ManagedMcpClientConfig {
  mcpServers: Record<string, McpServerEntry>;
}

/**
 * Produce an MCP client config snippet from a list of endpoints.
 *
 * Returns a dict in the standard `mcpServers` shape understood by Claude
 * Desktop, Cursor, and other MCP-aware clients. Each endpoint becomes one
 * server entry keyed `${name}.${kind}.${identifier}`.
 */
export function managedMcpClientConfig(
  options: ManagedMcpClientConfigOptions,
): ManagedMcpClientConfig {
  const parsed = ManagedMcpClientConfigOptionsSchema.parse(options);
  const name = parsed.name ?? 'databricks';
  const includeUnsupported = parsed.includeUnsupported ?? false;

  const servers: Record<string, McpServerEntry> = {};
  for (const ep of parsed.endpoints as ManagedMCPEndpoint[]) {
    if (ep.unsupported && !includeUnsupported) {
      continue;
    }
    const key = `${name}.${ep.kind}.${ep.identifier}`;
    const entry: McpServerEntry = { type: 'http', url: ep.url };
    if (ep.oauthScope) {
      entry.oauth_scope = ep.oauthScope;
    }
    if (ep.unsupported) {
      entry.unsupported = true;
    }
    servers[key] = entry;
  }
  return { mcpServers: servers };
}
